File: typescript/src/rates/FrankfurterRatesProvider.ts
import {RatesProvider} from "./RatesProvider";

export class FrankfurterRatesProvider implements RatesProvider {
    async getRateFrom(fromCurrency: string, toCurrency: string) {
        const host = 'api.frankfurter.dev';
        const resp = await fetch(`https://${host}/v1/latest?amount=1&from=${fromCurrency}&to=${toCurrency}`);
        const data = await resp.json();
        return data.rates[toCurrency];
    }
}

File: java/src/main/java/org/bank/bankaccount/FrankfurterRatesProvider.java
package org.bank.bankaccount;

import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

@Component
public class FrankfurterRatesProvider implements RatesProvider {

    public static final String FRANKFURTER_APP = "api.frankfurter.app";

    @Override
    public Double get(String from, String to) {
        String uri = "/latest?amount=1&from=" + from + "&to=" + to;
        var data = RestClient.builder().baseUrl("https://" + FRANKFURTER_APP).build().get().uri(uri).retrieve().body(FrankFurterJson.class);
        return data.rates().get(to);
    }
}
